sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala

```diff
@@ -76,7 +76,11 @@ trait QueryExecutionListener {
 class ExecutionListenerManager private[sql](session: SparkSession, loadExtensions: Boolean)
   extends Logging {
 
-  private val listenerBus = new ExecutionListenerBus(session)
+  private val listenerBus = new ExecutionListenerBus(session.sessionUUID)
+  session.sparkContext.listenerBus.addToSharedQueue(listenerBus)
+  session.sparkContext.cleaner.foreach { cleaner =>
```

Comment:
So, this fix is designed to be disabled when `spark.cleaner.referenceTracking=false`?

Reply:
To be honest, no. It fixes the memory leak so we generally always want it takes effect. And the `ContextCleaner` is an experienced framework that we can leverage to fix this issue elegantly without aggressive changes.

The config is introduced since Spark 1.0.0 and enabled by default. So I personally think it's acceptable to assume users run applications with cleaner enabled in most cases.

Reply:
I'm not against having that assumption. But, this doesn't fix memory leaks at all if `spark.cleaner.referenceTracking=false` technically, does it? We should be clear about the limitation of the PR.

Reply:
> But, this doesn't fix memory leaks at all if spark.cleaner.referenceTracking=false technically, does it? 

Yes, it doesn't.

Reply:
Let's mention it in the PR description to be clear.

Reply:
Updated.